eth_typing/abi.py

```diff
@@ -214,6 +200,20 @@ class ABIError(TypedDict):
     """Error input components."""
 
 
+class ABIFunctionInfo(TypedDict):
+    """
+    TypedDict to represent an `ABIFunction` with the function selector and
+    corresponding arguments.
+    """
+
+    abi: Union[ABIFunction, ABIError, ABIFallback, ABIReceive]
```

Comment:
I left a similar comment in the util with a similar name in *eth_utils*... should this `Type` be something else? This seems like it's more broad than just the `ABIFunction`.

Reply:
What do you think about the word "Interface"? Might be redundant as `ABIInterface`, but I think it could look something like:

```
ABIInterface = Union[ABIFunction, ABIConstructor, ABIFallback, ABIReceive, ABIError]
ABIElement = Union[ABIInterface, ABIEvent]
"""Base type for `ABIFunction` and `ABIEvent` types."""
ABI = Sequence[ABIElement]
```

Reply:
Hmm now that I'm looking more closely, `constructor`, `receive`, and `fallback` are all just specific "function" types. Maybe something like `ABIFunctionType = Union[ABIFunction, ABIConstructor, ABIFallback, ABIReceive]` and then in the `ABIFunction` class the `type` can be `Union[ABIFunctionType, ABIError]`?

Reply:
`ABIFunctionType` is much clearer, that works for me!

Reply:
Wait that already exists and I think it works almost exactly as I was imagining. We should confirm that but otherwise we can use the more general `ABIFunctionType` in places where they all make sense.

Reply:
In which case I think I'm good with `ABIFunctionInfo` since I do think they are all "function" types

Reply:
Thought on this some more. `ABIFunctionInfo` contains more than just the ABI (also includes the selector and arguments of the function). `ABIFunctionType` is used as the base type of any ABI fuction type that has common traits (`stateMutability`, `payable` and `constant`).

I think your comment was valid, the `abi` type of `ABIFunctionInfo` can be better. The docstrings can be improved. I also think the aliases can help clarify. What if we made a type alias for `Union[ABIFunction, ABIError, ABIFallback, ABIReceive]` called `ABICallable`? That would represent all function types except `ABIEvent` and `ABIError`. I think there are a couple places in the code that could also use that particular type alias.

Reply:
> ABIFunctionType is used as the base type of any ABI fuction type that has common traits (stateMutability, payable and constant).

I see. It looks like that was a recent addition too, right? Has that been released yet? If so, I like the idea of `ABICallable`. Otherwise, I lean toward calling the base class `ABIFunctionBase` (instead of the current `ABIFunctionType`) and using `ABIFunctionType` as the `Union` class we are talking about. Mostly because the `type` param [in some of the abi objects](https://www.quicknode.com/guides/ethereum-development/smart-contracts/what-is-an-abi#understanding-the-elements-of-abi) is what differentiates these different "types".

I think that's a pretty pedantic difference though so if we've already started down a path of `ABIFunctionType` having been released, we can definitely use the `ABICallable` as the `Union` class as you said.

Thoughts on that?

Reply:
Yeah the `ABIFunctionType` was released with the initial set of ABI types. `ABICallable` could potentially be in conflict with the builtin `Callable`, but maybe it would be better as `ABICallableType`?

Reply:
I think `ABICallableType` would just confuse me even more, having `ABIFunctionType` as the base class. I don't see any issues with `ABICallable`, should we just go with that?

Reply:
Yep I'm good with that!